src/GafferScene/IECoreScenePreview/PrimitiveAlgo.cpp

```diff
@@ -1084,141 +1196,115 @@ IECoreScene::PrimitivePtr PrimitiveAlgo::mergePrimitives(
+				);
 			}
 		},
 		tbb::auto_partitioner(),
 		taskGroupContext
 	);
 
 	// Set topology and other primitive type specific globals
+	result.finalize();
 
-	if( resultMesh )
-	{
-		if( resultMeshInterpolation == "" )
-		{
-			resultMeshInterpolation = "linear";
-		}
+	return result.result;
+}
 
-		resultMesh->setTopologyUnchecked( resultVerticesPerFaceData, resultVertexIdsData, totalInterpolation[ PrimitiveVariable::Vertex ], resultMeshInterpolation );
+// Why isn't this a static in RunTimeTyped?
+bool isInstanceOf( IECore::TypeId type, IECore::TypeId baseType )
+{
+	return type == baseType || RunTimeTyped::inheritsFrom( type, baseType );
+}
 
-		if( resultMeshInterpolateBound == "" )
-		{
-			resultMeshInterpolateBound = MeshPrimitive::interpolateBoundaryEdgeAndCorner;
-		}
-		resultMesh->setInterpolateBoundary( resultMeshInterpolateBound );
+} // namespace
+
+
+void PrimitiveAlgo::transformPrimitive(
+	IECoreScene::Primitive &primitive, Imath::M44f matrix,
+	const IECore::Canceller *canceller
+)
+{
+	if( matrix == Imath::M44f() )
+	{
+		// Early out for identity matrix
+		return;
+	}
+
+	Imath::M44f normalMatrix = normalTransform( matrix );
 
-		if( resultMeshFaceVaryingLI == "" )
+	for( const auto &[name, var] : primitive.variables )
+	{
+		Canceller::check( canceller );
+		V3fVectorData *vecVar = IECore::runTimeCast<V3fVectorData>( var.data.get() );
+		V3fData *vecConstVar = IECore::runTimeCast<V3fData>( var.data.get() );
+		if( !vecVar && !vecConstVar )
 		{
-			resultMeshFaceVaryingLI = MeshPrimitive::faceVaryingLinearInterpolationCornersPlus1;
+			continue;
 		}
-		resultMesh->setFaceVaryingLinearInterpolation( resultMeshFaceVaryingLI );
 
-		if( resultMeshTriangleSub == "" )
+		GeometricData::Interpretation interp = vecVar ? vecVar->getInterpretation() : vecConstVar->getInterpretation();
+		if( !(
+			interp == GeometricData::Interpretation::Point ||
+			interp == GeometricData::Interpretation::Vector ||
+			interp == GeometricData::Interpretation::Normal
+		) )
 		{
-			resultMeshTriangleSub = MeshPrimitive::triangleSubdivisionRuleCatmullClark;
+			continue;
 		}
-		resultMesh->setTriangleSubdivisionRule( resultMeshTriangleSub );
 
-		if( resultCornerIdsData )
+		if( vecVar )
 		{
-			resultMesh->setCorners( resultCornerIdsData.get(), resultCornerSharpnessesData.get() );
+			for( Imath::V3f &i : vecVar->writable() )
+			{
+				Canceller::check( canceller );
+				transformPrimVarValue( i, i, matrix, normalMatrix, interp );
+			}
 		}
-
-		if( resultCreaseLengthsData )
+		else
 		{
-			resultMesh->setCreases(
-				resultCreaseLengthsData.get(), resultCreaseIdsData.get(), resultCreaseSharpnessesData.get()
-			);
+			// Fairly weird corner case, but technically Constant primvars could need transforming too
+			transformPrimVarValue( vecConstVar->writable(), vecConstVar->writable(), matrix, normalMatrix, interp );
 		}
 	}
-	else if( resultCurves )
+}
+
+IECoreScene::PrimitivePtr PrimitiveAlgo::mergePrimitives(
```

Comment:
It's interesting that the implementation now basically switches on type at the very top, meaning that my preferred type-safe `mergeMeshes()/mergeCurves()/mergePoints()` API would no longer be an artificial construct on top of the main implementation. Simpler to leave things as they are for now though, rather than faff around separating things into different headers including a private one...